### ISSUE TITLE
robotics: Universal Embodiment v2 qualification program and dependency gates

### ISSUE THREAD

**Issue by Tristan-Stoltz-ERC** [Luminous-Dynamics/symthaea#1288]
This is the coordination issue for the current Universal Embodiment / world-model / HIL program. The individual issues are intentionally narrow; this issue defines the **implementation order and graduation gates** so higher layers do not build evidence or authority on unresolved lower-layer semantics.

## Program invariant

Do not equate:

- compile success with simulator qualification;
- simulator qualification with HIL;
- HIL with operational safety;
- semantic alignment with learned transfer;
- capability with authority;
- command publication with execution;
- a finite metric with trustworthy evidence.

Every layer should be usable at the strongest evidence level it has actually earned, while preserving compatibility paths separately.

### A1. #1279 — explicit health/reliability availability

Close synthetic-perfect evidence defaults before v2 safety/capability consumes them.

Graduation gate:

- missing health/reliability/identity is explicit;
- compatibility priors cannot masquerade as measured evidence;
- claim-bearing consumers can demand evidence class/freshness.

### A2. #1281 — metric semantic typing

Separate predictive residual, temporal novelty, task error, observation confidence, divergence.

Graduation gate:

- platform audit table exists;
- new-platform scaffold no longer generates false `prediction_error` semantics;
- v2 grounding only promotes temporal-model grounding from qualified predictive residual evidence.

### A3. #1253/#1254 — execution lineage

Reuse exact environment/backend/toolchain/commit lineage for every later qualification artifact.

No later HIL/simulator qualification report is authoritative without execution lineage.

### B1. #1284 — cognitive time vs plant time

This is a hard prerequisite for HIL and sequence-model evidence.

Graduation gate:

- cognitive tau, scheduler cadence, plant/model time, and transport latency are separate;
- actual simulator/device transition dt is observable;
- all `tau_factor` consumers are classified and the faster/slower convention is tested.

### B2. #1269 — recurrent snapshot/restore

Graduation gate:

- versioned latent-state snapshot/restore exists;
- checkpoint/model identity and latent identity are explicit;
- restored continuation reproduces the declared deterministic trace tolerance.

### C1. #1274 — core typed contract

This is the central convergence layer, not a new parallel robotics crate.

Minimum v2 primitives before downstream migration:

- extensible embodiment/profile identity hooks;
- typed observation frame;
- typed requested/projected/applied action stages;
- explicit time/clock references;
- calibration/frame/profile references;
- capability/evidence references;
- compatibility facade for `EmbodimentBridge` v1.

### C2. #1276 — readiness / plant safety / authority split

Must integrate at the action boundary, not as another post-hoc telemetry field.

Graduation gate:

- high Phi/Ψ alone cannot positively establish plant-safe full authority;
- missing required plant-safety evidence is restrictive;
- cognition may reduce goal-directed authority without disabling qualified local fallback;
- legacy MotorSafetyLevel is a downstream compatibility projection.

### C3. #1285 — provider/binding/lifecycle

Graduation gate:

- platform family, exact profile, backend provider, and bound instance are distinct;
- discovery has no actuation side effects;
- live-device binding can require runtime locators/calibration without faking genesis-only construction;
- third-party/provider IDs do not require a new central enum variant.

### D1. #1286 — interactive backend conformance

Use simple multirotor sim + MuJoCo as the first pair.

Graduation gate:

- requested vs actual dt receipt;
- truth vs observation separation;
- explicit backend/model/filter/context identity;
- reset/disturbance/mutation evidence;
- no dry-run engineering metric can be mistaken for live interactive qualification.

### D2. #1277 — multirotor semantic migration

Implement only after the relevant C1/B1 primitives are stable enough to avoid duplicate temporary schemas.

Graduation gate:

- `SemanticFlightFrameV1` and typed aggregate action profile;
- requested vs effective/applied wrench evidence;
- same semantic physical roles remain stable across simple-sim/MuJoCo;
- hidden derivative state is absent from universal state identity.

### E1. #1272 — typed residual/divergence

Can develop contracts in parallel, but simulator/HIL claims depend on Tracks A-D.

Graduation gate:

- residuals bind exact semantic role/contract/model/observation;
- uncertainty semantics are explicit and calibration-grade uncertainty is distinguished from heuristic scores;
- divergence policy is typed, fresh, reproducible, and platform-neutral.

### E2. #1255/#1256/#1257 — state repair, compact features, attribution

Complete before claiming useful action-sensitive or varied-state prediction.

### E3. #1271 — R4.8 sequence/rollout qualification

Depends on B1+B2 and the relevant model/evidence prerequisites.

Graduation gate:

- independent one-step, teacher-forced, conditioned, role-closed, and full-open-loop modes are distinct;
- episode/group leakage is prevented;
- every future/oracle input is declared;
- step horizons and elapsed plant seconds are both reported;
- recurrent-memory value is established against a reset/disabled ablation.

### F1. #1287 — ROS 2 adapter migration

Do not block static semantic work on a particular ROS distro, but do block execution claims on the matrix gates.

Prerequisites for claim-bearing ROS execution:

- C1 action/observation semantics;
- C2 safety/authority admission;
- B1 time/freshness semantics;
- C3 lifecycle/binding.

Graduation ladder per distro/RMW/profile:

`Unverified -> Compiles -> PubSubQualified -> ActionQualified -> SimIntegrationQualified -> HILQualified`

A topic publish receipt is not an execution receipt.

### F2. flight-controller edge (PX4/ArduPilot/MAVLink)

Audit first; create/implement only after identifying current code overlap. It should use the same v2 state/action/time/lifecycle contract rather than becoming a second flight ontology.

### F3. embedded/industrial edges

Later: micro-ROS/RTOS, CAN/CAN-FD, EtherCAT, OPC UA, etc. Do not add all protocol crates to core. Each is an edge/provider adapter against the same v2 semantics.

## Track G — cross-embodiment transfer

Do **not** open learned-transfer claims until at least humanoid + one actuated non-humanoid platform expose qualified semantic state/action contracts.

First candidate after #1277:

humanoid <-> multirotor shared body-root roles.

Required claim ladder:

1. shared physical role;
2. exact contract identity;
3. same-value representation identity;
4. pretrained/shared-model compatibility;
5. zero-shot predictive transfer;
6. fine-tuned transfer advantage;
7. controller/task transfer, if ever studied.

Each step requires its own baseline. Do not jump from (2) to (7).

## Track H — multiple-embodiment orchestration

Only after C1/C3 capability/provider identity is stable.

Goal:

`task requirements + available embodiment capabilities + runtime health + energy + authority + safety + evidence -> admissible delegation set`

Then planning chooses among admissible embodiments.

No central planner should infer authorization from capability discovery.

### Wave 0 — stop semantic leakage

- #1279
- #1281
- #1284 usage audit / type skeletons
- execution-lineage prerequisites

No behavior-wide refactor required yet.

### Wave 1 — v2 skeleton, compatibility-only

- minimal #1274 core types/traits behind no behavior change;
- legacy bridge adapter;
- typed availability/source/time/evidence records;
- unit tests only.

### Wave 2 — first concrete plant

- #1277 multirotor typed state/action;
- #1286 simple-sim receipts first;
- then MuJoCo.

Do not migrate the whole fleet simultaneously.

### Wave 3 — authority and runtime provider

- #1276 action admission/fallback separation;
- #1285 explicit provider/lifecycle;
- preserve legacy paths while measuring equivalence/regressions.

### Wave 4 — ROS simulator integration

- #1287 semantic ROS ingestion/output;
- simulator-backed ROS qualification;
- monitored trajectory action where bindings permit.

### Wave 5 — HIL

Only after simulator and transport gates are green.

HIL must record:

- exact hardware/profile/calibration;
- clock/transport timing;
- requested/projected/accepted/applied/measured action stages;
- health/watchdog/safety evidence;
- execution lineage;
- divergence/residual evidence.

### Wave 6 — second embodiment / transfer / fleet orchestration

Expand only once the first v2 path is boring and repeatable.

## CI/evidence policy

Each qualification PR should state:

- exact parent/head;
- exact feature set/devshell/toolchain;
- exact tests run;
- which evidence level it earns;
- explicit non-claims.

Prefer path-scoped gates for expensive robotics dependencies.

Never silently upgrade an adapter maturity state because adjacent code compiles.

## Compatibility policy

The migration should be additive first.

Do not break all 15+ existing embodiment implementations at once. Keep v1 behavior available, but mark v1-derived assumptions/proxies so new v2 evidence code cannot accidentally treat them as qualified facts.

## Definition of 'universal' for this program

Do not use universality as a completed implementation claim until the conformance suite has demonstrated at least:

- two materially different actuated embodiment classes;
- two interactive simulator backends for one platform;
- one ROS/runtime middleware edge;
- one embedded/flight-controller or comparable non-ROS edge;
- explicit time/frame/calibration/evidence semantics;
- provider lifecycle and runtime degradation;
- action authority/safety separation;
- reproducible replay/qualification.

Until then, 'universal' describes the architecture target, not verified ecosystem coverage.

**Comment by Tristan-Stoltz-ERC** [comment]
Program update — two new tracks discovered during the code audit:

### F2 now concrete: #1291 autopilot provider

PX4/ArduPilot/MAVLink should be implemented as a transport/product profile beneath the same v2 semantic flight contract, not as a second flight ontology. #1291 adds SITL-first qualification, explicit liveness/mode/ACK/time-sync/frame/security semantics, and keeps local flight controllers responsible for inner realtime loops.

This remains downstream of the relevant #1274/#1284/#1285/#1276/#1277 foundations.

### H prerequisite: #1293 authenticated capability claims

Before multi-embodiment delegation consumes swarm `CapabilityCard`s, separate content integrity, authenticated authorship, and evidence-supported capability. Current v1 card hash is an integrity digest, not a signature, and legacy `ReputationBridge::process_card()` counts card receipts as interactions.

Wave-0 migration seam is now Draft PR #1295 at exact head `fa17560d6c3cd814bb6d52c95e0626ec0b7b79f7`: it adds non-mutating `inspect_card()` plus explicit `record_successful_interaction()` while preserving legacy behavior. Exact-head CI is pending; do not promote/merge based on this comment.

**Comment by Tristan-Stoltz-ERC** [comment]
Wave-0 implementation update:

- Draft #1332 (`327a46b7141d51789b148529588bea3415c66840`) extracts the generic R3.10 causal physical-action stage + evidence-binding rule into core, without duplicating R3.1 physical-role semantics.
- Draft #1336 (`7bb99d86d0dd90119b3ecd7d07b23be070cea1b6`) stacks on #1332 and adds protocol-neutral transport/application delivery evidence, explicitly preventing ROS/DDS/MAVLink ACKs from being treated as backend-applied or measured physical execution.
- Draft #1312 remains the separate plant-time receipt layer.

The intended composition is now:

`physical action proposition -> transport/application delivery proposition -> backend-applied physical proposition -> plant timing/transition -> measured execution/observation`

These are independent evidence dimensions; do not collapse them into one monotonic 'execution status'. All remain draft pending exact-head qualification.

**Comment by Tristan-Stoltz-ERC** [comment]
Time/transport qualification update:

- #1332: generic physical action stages (`327a46b7141d51789b148529588bea3415c66840`)
- #1336: transport/application delivery states (`7bb99d86d0dd90119b3ecd7d07b23be070cea1b6`)
- #1339: correlated content-addressed timestamped delivery traces (`c39dd180950cbd34c0beaaabdd69a18865156fc9`)
- #1343: one-sample, evidence-bound directed cross-clock correspondence (`b882cdc022c9808e842a0fb263210d2a515d4bc0`)
- #1312 remains the separate physical plant-transition timing receipt.

The key invariant is now explicit in code: transport event ordering can span clock domains, but arithmetic across those clocks fails closed unless a separate clock-alignment layer earns comparability. A single #1343 sample still does not establish a general clock transform or drift model.

All listed PRs remain draft pending exact-head execution evidence.

**Comment by Tristan-Stoltz-ERC** [comment]
Program refinement from the Reach/Grasp qualification audit: track #1398 as the shared qualification-stage envelope/transition contract.

The important proposition split is:

```text
Simulation qualification
    != HIL qualification
    != Physical qualification
    != Operational authority
```

The first three should share an embodiment-neutral stage vocabulary and explicit transition policy, but `Operational` must remain outside that qualification enum. Stage identity says where/how evidence was earned; it is not a trust/authority ordering and should not derive `Ord`.

Reach provides the compatibility proof because its strongest manifest path already checks exact Simulation/HIL/Physical purpose identities before operational composition. Grasp #1390 should be the first native high-assurance consumer once its provenance/census/geometry theorem is qualified.

#1398 must compose with the existing v2 dimensions rather than collapse them: #1332 physical action stages, #1336/#1339 transport delivery, #1312 plant timing, #1343/#1348 clock correspondence, #1276 live plant-safety/authority separation. None of those is itself a qualification-stage transition.

**Comment by Tristan-Stoltz-ERC** [comment]
Grasp reference ladder update under the Universal Embodiment program:

- #1398 defines generic Simulation/HIL/Physical stage envelopes + explicit transition policy, with Operational deliberately excluded from the stage enum.
- #1399 defines Grasp HIL as NEW evidence: requested/safety-projected action, transport/application delivery, backend application, independent measured execution, HIL reference instrumentation, timing/watchdog/fault evidence.
- #1400 defines bounded Physical Grasp on the actual robot under a controlled no-human-contact v1 test envelope, adding robot-instance/configuration, health/calibration, thermal/load/duty-cycle, fallback/E-stop and whole-body stability evidence.
- #1401 defines separate action-scoped Operational admission over current Physical qualification + mission scope + live authority + plant safety + one-use transactional effect admission.

This sharpens the Wave 5+ invariant: HIL and Physical are evidence stages; Operational is a live authorization/safety composition, not a higher evidence-stage ordinal. Use this reference pattern when Lift/Carry/multirotor/provider paths mature rather than creating parallel ladders.

**Comment by Tristan-Stoltz-ERC** [comment]
Program update — qualification-stage scope and maintenance semantics are now separated explicitly.

- #1398: generic Simulation/HIL/Physical stage identity + transition semantics, with Operational authority deliberately outside the stage enum.
- #1413: exact qualification applicability binding. Evidence fidelity and applicability are orthogonal; a later stage may be higher-fidelity but narrower-scope. No predecessor-only region is silently upgraded, capability declarations cannot widen qualification, and future Operational admission must prove the requested action/scope lies inside the current Physical applicability theorem.
- #1414: conservative change-impact / delta-requalification theorem. Changed subjects start new lineages by default; evidence reuse must be earned claim-by-claim through exact subject snapshots, semantic change sets, dependency profiles, selection-complete delta campaigns, and explicit applicability relations. Old artifacts remain immutable historical evidence.

This should become a generic v2 program invariant:

```text
qualification stage
    != applicability scope
    != current operational authority
```

and:

```text
higher-fidelity evidence
    != wider applicability
changed version
    != preserved qualification
```

Do not implement #1413/#1414 as large speculative stacks while Actions are saturated. First qualify the small #1398 stage/envelope substrate, then add applicability identity, then change-impact composition.

**Comment by Tristan-Stoltz-ERC** [comment]
Adaptive-control hardening added as #1430.

Program rule to carry forward: `learning/adaptation != authority to replace the currently qualified controller`. Persistent controller/policy mutation creates a new candidate by default and routes through #1414 stage-specific impact/requalification. Only an explicitly qualified bounded adaptation law may evolve declared state/parameters inside its exact applicability envelope, with transition receipts and an external immutable guard.

Keep epistemic/body-schema adaptation, ephemeral qualified control state, and persistent policy mutation as separate propositions. This should apply across humanoid and later embodiments rather than becoming Grasp-specific.

**Comment by Tristan-Stoltz-ERC** [comment]
Physical lifecycle hardening added as #1431.

Universal Embodiment should treat Physical qualification as bounded historical evidence plus explicit continued-qualification obligations where the hardware/profile requires them. Calibration/inspection/service/cycle/condition surveillance is separate from fresh per-action plant safety and from #1424 counterexample restriction.

This prevents a once-green Physical campaign from becoming an indefinite capability certificate merely because the source/controller bytes did not change.
